icepyx/core/visualization.py

```diff
@@ -0,0 +1,340 @@
+                    lat=(["rgt", "cycle", "bbox", "index"], lat),
+                    lon=(["rgt", "cycle", "bbox", "index"], lon)),
+                coords=dict(
+                    rgt=rgt_no,
+                    cycle=cycle_no,
+                    index=index,
+                    bbox=bbox_no),
+                attrs=dict(description="ICESat-2 Elevation"))
+
+            return OA_ds
+
+    def parallel_request_OA(self):
+        """
+        Requesting elevation data from OpenAltimetry API in parallel
+        Now only supports OA_Products ['ATL06','ATL07','ATL08','ATL10','ATL12','ATL13']
+
+        For ATL03 Photon Data, OA only supports single date request
+        according to: https://openaltimetry.org/data/swagger-ui/#/Public/getATL08DataByDate,
+        with geospatial limitation of 1 degree lat/lon
+
+        :return: all requested ICESat-2 elevation in xarray dataset
+        """
+        print('Generating urls')
+
+        # generate parameter lists for OA requesting
+        OA_para_list = self.generate_OA_parameters()
+
+        print('Sending request to OpenAltimetry, please wait...')
+
+        # Parallel processing
+        requested_OA_data = []
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=len(OA_para_list)) as executor:
+            parallel_OA_data = {executor.submit(self.request_OA_data, para): para for para in OA_para_list}
+
+            for future in tqdm(
+                    iterable=concurrent.futures.as_completed(parallel_OA_data),
+                    total=len(parallel_OA_data),
+            ):
+                r = future.result()
+                if r is not None:
+                    requested_OA_data.append(r)
+
+        requested_OA_data_filtered = list(filter(None, requested_OA_data))
+
+        if not requested_OA_data_filtered:
+            return
+        else:
+            OA_data_ds = xr.merge(requested_OA_data_filtered)
+            return OA_data_ds
+
+    def visualize_extent(self):
+        """Show bounding box extent set by users"""
+        region = ipx.Query(self.product, self.bbox, self.date_range)
+        gdf = geospatial.geodataframe(region.extent_type, region._spat_extent)
+        line_geoms = Polygon(gdf['geometry'][0]).boundary
+        bbox_poly = gv.Path(line_geoms).opts(color="red", line_color="red")
+        tile = gv.tile_sources.EsriImagery.opts(width=500, height=500)
+        return tile * bbox_poly
+
+    def visualize_elevation(self):
+        """
+        Visualize elevation requested from OpenAltimetry API using datashader based on cycles
+        https://holoviz.org/tutorial/Large_Data.html
+        """
+
+        if self.product in ['ATL06', 'ATL07', 'ATL08', 'ATL10', 'ATL12', 'ATL13']:
+
+            OA_ds = self.parallel_request_OA()
+
+            if OA_ds is None:
+                print('No data')
+
+            else:
+                ddf = OA_ds.to_dask_dataframe().astype({'lat': 'float', 'lon': 'float','elevation': 'float'})
```

Comment:
Actually, I'm wondering if it's better to put everything in a dask.DataFrame directly, instead of going through an xarray.Dataset (though I imagine the xarray.Dataset format would be useful for storing attributes). What do you think?

```python
dask.dataframe.from_array(x=OA_array, columns=["lat", "lon", "elevation"])
```

```
Dask DataFrame Structure:
	lat	lon	elevation
npartitions=1			
0	float64	float64	float64
2831	...	...	...
Dask Name: from_array, 1 tasks
```

Reply:
Agree, `dask.dataframe` is better as in the `visualization_elevation` function I converted `xarray.dataset` into a `dask dataframe` anyway so really no need to generate xarray.dataset here.